### ISSUE TITLE
Nothing tells you before a run which of your 275 settings your backend will actually read — and it cannot be inferred, only declared

### ISSUE THREAD

**Issue by Srinivasan8888** [MakazhanAlpamys/Soup#755]
## The problem, stated as the user meets it

You write a config, it validates clean, training runs, and some of what you asked for
never happened. You find out afterwards — from a loss curve, or not at all.

That has been filed one field at a time: #683 (`train_on_responses_only` dropped on MLX),
#686 (warmup, scheduler, weight decay, optimizer choice — four in one backend), #745
(LoRA+ wired in 3 trainers, silently ignored by 13), #749 (`max_grad_norm` honoured by
sixteen transformers trainers, dropped on MLX). #748 counts 36 issues in this family by
tracker search.

`TrainingConfig` declares **237** fields and `DataConfig` **38** — 275 settings, and
nothing tells you before a run which of the ones *you set* will actually be read on the
task and backend *you chose*.

## What already exists, and why it is the right shape

`trainer/mlx_sft.py:139-179` builds an `unsupported` list and prints

```
MLX backend ignores: GaLore, Ring Attention, training.seed (MLX seeds through mx.random), ...
```

That mechanism is proven and already user-facing. Its limits are the whole proposal:

- **nine entries, one file, one backend.** Nothing equivalent exists for the sixteen
  transformers trainers, where #745's thirteen-trainer LoRA+ gap lives;
- **hand-maintained, and it drifted.** `max_grad_norm` was not in that list while MLX was
  dropping it — that is #749, found by reading code, not by any check;
- **it fires during the run**, not before it.

## Three implementations I measured and rejected

Recording these because each is the obvious answer and each is wrong, and a reviewer
would rightly ask why I did not just do them.

**1. Static reachability over the import graph — detects nothing.** Count a field as
reachable from a trainer if any module in its transitive `soup_cli` import closure reads
it as an attribute or names it as an exact string. Every trainer's closure is ~150
modules, i.e. most of the package, so almost everything looks reachable. Against four
gaps that are independently known to be real:

| field | flagged unreachable from `mlx_sft`? | ground truth |
|---|---|---|
| `max_grad_norm` | no | MLX gap (#749) |
| `warmup_steps` | no | MLX gap (#686) |
| `weight_decay` | no | MLX gap (#686) |
| `loraplus_lr_ratio` | no | not wired (#745) |

**Zero of four.** The set of fields it says MLX lacks but SFT has is empty.

**2. Static reads of the trainer module alone — false positives that would get it
deleted.** It does find the four above, but `sft.py` does not itself read
`train_on_responses_only` (that lives in `data/sft_format.py:55`) or `loraplus_lr_ratio`
(in the LoRA+ helper from #738), so it would report two gaps the transformers path does
not have. A checker that cries wolf about masking is one nobody reads twice.

**3. Runtime tracing under `--dry-run` — observes nothing relevant.** A config proxy
recording `__getattr__` is accurate by construction, but `--dry-run` raises `typer.Exit`
at `commands/train.py:1255` after validating data and **before the trainer wrapper is
constructed**, so no trainer field read ever happens. Making it go further means loading
a model, which a pre-flight check cannot do.

The honest conclusion: **which fields a backend honours cannot be inferred; it has to be
declared.** Inference is exactly what I tried three ways and could not make trustworthy.

## Proposal

**1. A declared support registry.** Promote the `unsupported` list out of `mlx_sft.py`
into one reviewed table keyed by (task, backend), covering the transformers trainers as
well as MLX. Each entry names a field, a status (`honoured` / `ignored` / `rejected`) and
a reason, and — where one exists — an issue number. The MLX nine and #745's thirteen
seed it.

**2. `soup doctor --config soup.yaml`.** `doctor` today is 16 functions of environment
diagnostics and reads no config at all (`grep -c "load_config\|soup.yaml"` → 0). Add a
config leg that loads the user's config, diffs the fields they actually set against the
registry for their task and backend, and reports only what affects them:

```
Config check — task=sft backend=mlx
  12 settings, 9 honoured

  ignored on this backend:
    training.max_grad_norm      MLX clips nowhere (#749)
    training.warmup_steps       MLX uses its own schedule (#686)
    data.mask_history           declared, no consumer anywhere (#748)
```

**3. A guard so the registry cannot drift.** Narrow enough to be trustworthy, which the
measurements above define: every registry entry names a real field on the model; a field
marked `ignored` for a backend must not be read by that backend's trainer module or its
declared helper modules; the helper list is explicit, because that is precisely where
implementation 2 went wrong.

### Acceptance criteria

1. `soup doctor --config` names `training.max_grad_norm` as ignored for `task=sft
   backend=mlx` on today's `main`, and stops naming it when #750 lands — demonstrated
   both ways, not described.
2. It reports only fields the user actually set, not all 275.
3. A field marked `honoured` that no code reads fails the guard.
4. A registry entry naming a field that no longer exists fails the guard.
5. The transformers path is not reported as missing `train_on_responses_only` or
   `loraplus_lr_ratio` — the false positives from implementation 2, each with its own
   test.
6. Runs on every CI cell — no GPU, no MLX, no network, no model download.

### Scope limits

- The registry is **declared, not derived**. It is only as right as its review, and I
  would rather be honest that this is a maintained table than pretend a walker can
  infer it. The guard bounds the drift; it cannot originate the truth.
- Per-*trainer* granularity, not per-code-path: a field honoured on one branch of a
  trainer and not another reads as honoured.
- I would ship the registry plus `doctor --config` for `sft` and MLX first, with the
  other tasks seeded from #745's existing `_LORAPLUS_NOT_IMPLEMENTED` list, rather than
  claiming all 275 × 18 cells are reviewed on day one.

Taking this, on the understanding that the registry's shape is your call and I will
rework it on your word — this is the same architectural decision point as #745, and I
would rather show you a small honest slice than a large inferred one. No hardware needed
beyond a laptop; no Apple Silicon required for the guard or the doctor leg.

**Comment by Srinivasan8888** [comment]
**Correction to my own body, found by this repo's own unknown-key warning rather than by review.**

The rejection table names `training.warmup_steps` as one of the four known MLX gaps. **There is no such field.** Writing it into a config produces exactly what #628 was built to produce:

```
Warning: unknown config key 'training.warmup_steps' - did you mean 'warmup_auto'
or 'warmup_ratio'? Not applied.
```

The real fields are `training.warmup_ratio` (`schema.py:1096`), `weight_decay` (`:1097`), `max_grad_norm` (`:1098`), `optimizer` (`:1164`) and `scheduler` (`:1233`). I took "warmup" from #686's prose title and did not check it against the schema before writing the table.

**The measurement is unchanged.** Re-run with the corrected names, implementation 1 still detects none of them — and I added `optimizer` while I was there, so it is now zero of five rather than zero of four:

| field | flagged unreachable from `mlx_sft`? | read by `sft.py`? |
|---|---|---|
| `max_grad_norm` | no | yes |
| `warmup_ratio` | no | yes |
| `weight_decay` | no | yes |
| `optimizer` | no | yes |
| `loraplus_lr_ratio` | no | no (#738 helper) |

Every one of the five is read by `trainer/sft.py` and by no line of `trainer/mlx_sft.py` on `254351e1`, which is the gap the feature is meant to surface — and which the transitive-closure walker reports as fully covered.

Flagging it because the acceptance criteria are written to be re-run, and a reviewer copying `warmup_steps` out of the table would hit the warning above and reasonably wonder what else was unchecked.

**Comment by MakazhanAlpamys** [comment]
Closed by #756, contributed by @Srinivasan8888, who filed this issue and then built it.

`soup doctor --config soup.yaml` now lists the settings your config actually writes that your task and backend do not read, each with the reason and, where one exists, the issue that recorded it. It exits 2 when the config cannot be read, so it can gate CI.

The scope is deliberately narrow for now: `task=sft` on `backend=mlx`, seven entries, all backed by warnings `trainer/mlx_sft.py` already emits at runtime. Every other task/backend pair reports nothing rather than guessing. A guard keeps the table honest in both directions, so adding coverage for another pair means adding declared entries, not trusting inference.

Try it on `main`: `soup doctor --config soup.yaml`
